### DISCUSSION TITLE
Is there a way to add custom logger for API calls made through Semantic Kernel?

### DISCUSSION THREAD

**Discussion by prajwalpyakurelMS** [microsoft/semantic-kernel#6714] [Q&A] [answered]
We would like to have our custom logger log events for api calls made through semantic kernel (OpenAI, plugins/tools, etc). 
Is there a way to achieve this using filter or something similar?
We use C#.

We do have dependency calls logged in our infrastructure, but we'd like to have custom logging in the code so we can have more control over it. 

Thank you!

**Comment by dmytrostruk** [comment]
@prajwalpyakurelMS Could you please specify which SK language you are using? This will help us to provide related examples. Thanks!

**Reply by prajwalpyakurelMS** [reply]
@dmytrostruk We are using C#. We're looking for api call data like url, status, etc

**Reply by dmytrostruk** [reply]
@prajwalpyakurelMS If you want custom logging on API level, you can inject custom `HttpClient` and extend `DelegatingHandler` to log information you need. 

This is an example how to extend `DelegatingHandler` with logging:
https://github.com/microsoft/semantic-kernel/blob/1cc6464860b84e1d781db7ac4a6b047dd1751b40/dotnet/src/InternalUtilities/samples/InternalUtilities/BaseTest.cs#L103-L130

Then you can initialize `HttpClient` with `DelegatingHandler` and pass your custom `HttpClient` to AI connector of your choice:
https://github.com/microsoft/semantic-kernel/blob/1cc6464860b84e1d781db7ac4a6b047dd1751b40/dotnet/samples/Concepts/ChatPrompts/SafeChatPrompts.cs#L16-L25

Let me know if this is helpful for your scenario.

**Reply by prajwalpyakurelMS** [reply]
Thanks for the example. Can you confirm that we will be using SendAsync only so we only need to override it? Because if you use other methods like GetAsync, PutAsync, etc. we might have to override other methods as well.

**Reply by dmytrostruk** [reply]
@prajwalpyakurelMS For our connectors, we try to use existing SDKs (e.g. Azure .NET SDK), so this functionality is implemented on their side. But usually overriding `SendAsync` method should be enough. We use the same approach for unit testing and mocking.

**Reply by prajwalpyakurelMS** [reply]
Ok thanks a lot! I have a followup:

Can I simply override HttpClientHandler like this in that case? Is there an advantage to doing the way you have described above?

Reason is I think this could apply to other use cases beyond kernel as well. 

```
public class CustomHttpClientHandler : HttpClientHandler
{
   private readonly ITestOutputHelper _output = output; 
    protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
    {
         // add logging

          var responseContent  = await base.SendAsync(request, cancellationToken);
          this._output.WriteLine(responseContent); 
         return responseContent;
    } 
}
```
Source: https://medium.com/@josemarmolejos/extend-httpclient-calls-and-add-authentication-mocking-and-other-goodies-838de768dbfb

**Reply by dmytrostruk** [reply, marked as answer]
@prajwalpyakurelMS You can use `HttpClientHandler` and it should work correctly, but I think for your scenario `DelegatingHandler` is a recommended way, because you can create a chain of delegating handlers where each handler could be responsible for logging, authentication. retry policies etc, while `HttpClientHandler` sits at the end of this chain and performs actual HTTP communication. This applies to HTTP communication in ASP.NET Web API, but you can use the same strategy in your app:
https://learn.microsoft.com/en-us/aspnet/web-api/overview/advanced/http-message-handlers
